--- /opt/openemr-7.0.3/openemr/src/Common/Auth/AuthUtils.php.orig	2025-03-23 06:57:52.000000000 +0000
+++ /opt/openemr-7.0.3/openemr/src/Common/Auth/AuthUtils.php	2025-11-10 05:39:45.703649590 +0000
@@ -44,6 +44,7 @@
 use OpenEMR\Common\Acl\AclExtended;
 use OpenEMR\Common\Acl\AclMain;
 use OpenEMR\Common\Auth\AuthHash;
+use OpenEMR\Common\Auth\AuthSandstorm;
 use OpenEMR\Common\Logging\EventAuditLogger;
 use OpenEMR\Common\Utils\RandomGenUtils;
 use OpenEMR\Services\UserService;
@@ -505,11 +506,10 @@
      * @param type $newPwd          the new password for the target user
      *                              - password is passed by reference so that it can be "cleared out" as soon as we are done with it.
      * @param type $create          Are we creating a new user or
-     * @param type $insert_sql      SQL to run to create the row in "users" (and generate a new id) when needed.
-     * @param type $new_username    The username for a new user
+     * @param type $userFormValues  Values used to create the user, if necessary
      * @return boolean              Was the password successfully updated/created? If false, then $this->errorMessage will tell you why it failed.
      */
-    public function updatePassword($activeUser, $targetUser, &$currentPwd, &$newPwd, $create = false, $insert_sql = "", $new_username = null)
+    public function updatePassword($activeUser, $targetUser, &$currentPwd, &$newPwd, $create = false, $userFormValues = [])
     {
         if (empty($activeUser) || empty($currentPwd)) {
             $this->errorMessage = xl("Password update error!");
@@ -596,9 +596,10 @@
         // (note that in this case, a random password is prepared for the new user below that is stored in OpenEMR
         //  and used only for session confirmations; the primary authentication for the new user will be done via
         //  LDAP)
-        $ldapDummyPassword = false;
-        if ($create && ($userInfo === false) && (!empty($new_username)) && (self::useActiveDirectory($new_username))) {
-            $ldapDummyPassword = true;
+        $validatePassword = true;
+        $newUsername = isset($userFormValues['username']) ? $userFormValues['username'] : '';
+        if ($create && ($userInfo === false) && (!empty($newUsername)) && (self::useActiveDirectory($newUsername))) {
+            $validatePassword = false;
             $newPwd = RandomGenUtils::produceRandomString(32, "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789");
             if (empty($newPwd)) {
                 // Something is seriously wrong with the random generator
@@ -616,19 +617,19 @@
         }
 
         // Ensure password is long enough, if this option is on (note LDAP skips this)
-        if ((!$ldapDummyPassword) && (!$this->testMinimumPasswordLength($newPwd))) {
+        if (($validatePassword) && (!$this->testMinimumPasswordLength($newPwd))) {
             $this->clearFromMemory($newPwd);
             return false;
         }
 
         // Ensure password is not too long (note LDAP skips this)
-        if ((!$ldapDummyPassword) && (!$this->testMaximumPasswordLength($newPwd))) {
+        if (($validatePassword) && (!$this->testMaximumPasswordLength($newPwd))) {
             $this->clearFromMemory($newPwd);
             return false;
         }
 
         // Ensure new password is strong enough, if this option is on (note LDAP skips this)
-        if ((!$ldapDummyPassword) && (!$this->testPasswordStrength($newPwd))) {
+        if (($validatePassword) && (!$this->testPasswordStrength($newPwd))) {
             $this->clearFromMemory($newPwd);
             return false;
         }
@@ -637,18 +638,7 @@
             // No userInfo means a new user
             // In these cases don't worry about password history
             if ($create) {
-                if (empty($new_username)) {
-                    $this->errorMessage = xl("Password update error!");
-                    $this->clearFromMemory($newPwd);
-                    return false;
-                }
-                // Collect the new user id from the users table
-                privStatement($insert_sql, array());
-                $getUserID = "SELECT `id`" .
-                    " FROM `users`" .
-                    " WHERE BINARY `username` = ?";
-                $user_id = privQuery($getUserID, [$new_username]);
-                if (empty($user_id) || empty($user_id['id'])) {
+                if (empty($newUsername)) {
                     $this->errorMessage = xl("Password update error!");
                     $this->clearFromMemory($newPwd);
                     return false;
@@ -660,11 +650,13 @@
                     error_log('OpenEMR Error : OpenEMR is not working because unable to create a hash.');
                     die("OpenEMR Error : OpenEMR is not working because unable to create a hash.");
                 }
-                // Store the new user credentials
-                $passwordSQL = "INSERT INTO `users_secure`" .
-                    " (`id`,`username`,`password`,`last_update_password`)" .
-                    " VALUES (?,?,?,NOW()) ";
-                privStatement($passwordSQL, [$user_id['id'], $new_username, $hash]);
+
+                $userId = UserService::createUserWithUserFormValuesAndHash($userFormValues, $hash);
+                if (empty($userId)) {
+                    $this->errorMessage = xl("Password update error!");
+                    $this->clearFromMemory($newPwd);
+                    return false;
+                }
             } else {
                 $this->errorMessage = xl("Missing user credentials") . ":" . $targetUser;
                 $this->clearFromMemory($newPwd);
@@ -1371,6 +1363,116 @@
     }
 
     /**
+     * Authenticate a user with the HTTP headers provided by Sandstorm.
+     * @param AuthSandstorm $auth_sandstorm
+     */
+    public static function verifySandstormAuth(AuthSandstorm $authSandstorm): void
+    {
+        $event = 'login';
+        $logPrefix = 'verifySandstormAuth';
+        // There is no meaningful way to use Open-EMR as an anonymous user.
+        // Stop processing if the user is anonymous.
+        $authSandstorm->dieIfAnonymous();
+
+        $queryResponse = sqlQuery(
+            "SELECT `user_id` " .
+            "FROM `sandstorm_users` " .
+            "WHERE `sandstorm_user_id` = ?",
+            [$authSandstorm->getUserIdAsBinary()]
+        );
+        if ($queryResponse === FALSE) {
+            // If Sandstorm user does not have an Open-EMR user, create one.
+            $newUsername = newUsernameWithPreferredHandle($authSandstorm->getPreferredHandle());
+            $userFormValues = UserService::buildUserFormValuesArray(
+                $newUsername,
+                $authSandstorm->getDisplayName(),
+                '', // middleName
+                '', // lastName
+                '', // suffix
+                NULL, // googleSigninEmail
+                '', // valedictory
+                '', // federalTaxId
+                '', // stateLicenseNumber
+                '', // erxRole
+                '', // physicianType
+                '', // mainMenuRole
+                '', // patientMenuRole
+                '', // erxPrId
+                '', // authorized
+                '', // info
+                '', // federalDrugId
+                '', // upin
+                '', // npi
+                '', // taxonomy
+                '', // facilityId
+                '', // billingFacilityId
+                '', // specialty
+                '', // seeAuth
+                '', // defaultWarehouse
+                '', // irnPool
+                0,  // calendar
+                0,  // portalUser
+                0,  // supervisorId
+            );
+
+            // Step 1. Add the user to the `users` table.
+            // Step 2. Add the user to the `users_secure` table.
+            sqlBeginTrans();
+            // No password should match a hash of '!'.
+            $oemrUserId = UserService::createUserWithUserFormValuesAndHash($userFormValues, '!');
+
+            // Step 3. Add the user ACL objects.
+            $groups = $authSandstorm->getAroGroupsWithSandstormPermissions();
+            AclExtended::addUserAros($userFormValues['username'], $groups);
+
+            // Step 4. Add the user to a group.
+            $authGroup = 'Default';
+            sqlInsert(
+                "INSERT INTO `groups` (name, user) " .
+                "VALUES (?, ?)",
+                [$authGroup, $userFormValues['username']]
+            );
+
+            // Step 5. Associate our new Open-EMR user with the current Sandstorm user.
+            sqlInsert(
+                "INSERT INTO `sandstorm_users` (sandstorm_user_id, user_id) " .
+                "VALUES (?, ?)",
+                [$authSandstorm->getUserIdAsBinary(), $oemrUserId]
+            );
+            sqlCommitTrans();
+
+            // TODO: Fire an event since we created a new user.
+        } else if (!empty($queryResponse)) {
+            $oemrUserId = $queryResponse['user_id'];
+        }
+
+        $oemrUserInfo = privQuery(
+            "SELECT `id`, `username`, `authorized`, `see_auth`, `active` " .
+            "FROM `users` " .
+            "WHERE `id` = ?",
+            [$oemrUserId]
+        );
+        if ($oemrUserInfo == FALSE) {
+            $message = $logPrefix. ": " . " Sandstorm user \"" . $authSandstorm->getDisplayName() . "\" (" . $authSandstorm->getPreferredHandle() . ") with Sandstorm user ID \"" . $authSandstorm->getUserIdAsString() . "\" and Open-EMR user ID \"" . $oemrUserId . "\" was not found in the `users` table.";
+            EventAuditLogger::instance()->newEvent($event, '', '', 0, $message);
+            $authSandstorm->dieWithMessage($message);
+        }
+        if (empty($authGroup)) {
+            $groupResult = sqlQuery(
+                "SELECT `name` " .
+                "FROM `groups` " .
+                "WHERE `user` = ?",
+                [$oemrUserInfo['username']]
+            );
+            if ($groupResult !== FALSE) {
+                $authGroup = $groupResult['name'];
+            }
+        }
+
+        AuthUtils::setUserSessionVariables($oemrUserInfo['username'], 'SANDSTORM_USER_NO_PASSWORD', $oemrUserInfo, $authGroup);
+    }
+
+    /**
      * Given an associative array representing the user, set the session variables
      * @param $username
      * @param $hash
@@ -1391,3 +1493,29 @@
         }
     }
 }
+
+function newUsernameWithPreferredHandle(string $handle): string {
+
+    $alreadyExists = true;
+    $number = 1;
+    $testedHandle = $handle;
+    $tooManyTries = 100;
+
+    while ($alreadyExists) {
+        $exists = sqlQuery(
+            "SELECT 1 " .
+            "FROM `users` " .
+            "WHERE `username` = ?",
+            [$testedHandle]
+        );
+        if ($exists === FALSE) {
+            break;
+        }
+        $testedHandle = $handle . "-" . $number;
+        $number += 1;
+        if ($number >= $tooManyTries) {
+            die("Giving up after trying to find an unused username " . $number . " times (handle=" . $handle . ")");
+        }
+    }
+    return $testedHandle;
+}
