--- /opt/openemr-7.0.3/openemr/library/auth.inc.php.orig	2025-10-06 17:02:20.086799373 +0000
+++ /opt/openemr-7.0.3/openemr/library/auth.inc.php	2025-10-20 10:58:58.294912963 +0000
@@ -17,10 +17,23 @@
  */
 
 use OpenEMR\Common\Auth\AuthUtils;
+use OpenEMR\Common\Auth\AuthSandstorm;
 use OpenEMR\Common\Logging\EventAuditLogger;
 use OpenEMR\Common\Session\SessionTracker;
 use OpenEMR\Services\UserService;
 
+if (getenv('SANDSTORM') == '1') {
+    // TODO: Where will we set the language if we have no login screen?
+    $_SESSION['language_choice'] = 1;
+    $_SESSION['language_direction'] = getLanguageDir($_SESSION['language_choice']);
+
+    $authSandstorm = new AuthSandstorm();
+    $authSandstorm->dieIfAnonymous();
+    $login_success = true;
+
+    return;
+}
+
 $incoming_site_id = '';
 // This is the conditional that ensures that the submission has the required parameters to attempt a login
 if (
