--- /opt/openemr-7.0.3/openemr/templates/login/layouts/vertical_box.html.twig.orig	2025-11-05 18:07:23.554202778 +0000
+++ /opt/openemr-7.0.3/openemr/templates/login/layouts/vertical_box.html.twig	2025-11-05 18:08:26.176202858 +0000
@@ -14,7 +14,7 @@
 
 {% block content %}
 <div class="vh-100 p-5 d-flex flex-column justify-content-center">
-    <form method="POST" id="login_form" autocomplete="off" action="../main/main_screen.php?auth=login&site={{ siteID|attr_url }}" target="_top" name="login_form">
+    <form method="POST" id="login_form" autocomplete="off" action="../main/main_screen.php?auth=login&site={{ siteID|attr_url }}" name="login_form">
         <div class="shadow-lg m-0 my-sm-5 login-row-center">
 
             <div class="py-2 px-2 py-md-3 px-md-5 box d-flex flex-column justify-content-center w-100">
