--- /dev/null	2025-11-10 05:13:38.390906557 +0000
+++ /opt/openemr-7.0.3/openemr/src/Common/Auth/AuthSandstorm.php	2025-11-10 11:29:05.325963868 +0000
@@ -0,0 +1,232 @@
+<?php
+
+/**
+ * AuthSandstorm class.
+ *
+ *   Support for Sandstorm's authentication model
+ *
+ * @package   Open-EMR for Sandstorm
+ * @link      https://github.com/sandstorm-org/openemr-sandstorm
+ * @author    Marcus Yu <saguiau4417@gmail.com>
+ * @author    Troy J. Farrell <troy@entheossoft.com>
+ * @copyright Copyright (c) 2025 Marcus Yu, Troy J. Farrell
+ * @license   https://github.com/openemr/openemr/blob/master/LICENSE GNU General Public License 3
+ */
+
+namespace OpenEMR\Common\Auth;
+
+class AuthSandstorm
+{
+    /**
+     * @var string
+     */
+    private $displayName;
+    /**
+     * @var string[]
+     */
+    private $permissions;
+    /**
+     * @var string
+     */
+    private $pictureUrl;
+    /**
+     * @var string
+     */
+    private $preferredHandle;
+    /**
+     * @var string
+     */
+    private $pronouns;
+    /**
+     * @var string
+     */
+    private $sessionType;
+    /**
+     * @var string
+     */
+    private $sessionId;
+    /**
+     * @var string
+     */
+    private $tabId;
+    /**
+     * @var string
+     */
+    private $userIdString;
+    /**
+     * @var string
+     */
+    private $userIdBinary;
+
+
+    public function __construct() {
+        $this->populateWithHeaders();
+    }
+
+    /**
+     * Die if the user is anonymous
+     */
+    public function dieIfAnonymous() {
+        if (empty($this->userIdString)) {
+            $this->dieWithMessage('Please Log In with Sandstorm to use Open-EMR');
+        }
+    }
+
+    /**
+     * Die with a message
+     * @param string $message
+     */
+    public function dieWithMessage($message) {
+        $style_block = '<div style="position: absolute;left: 25%;display: flex;justify-content: center;width: 50%;height: 100%;align-items: center;background-color: #f8f9fa;">';
+        $style_text = '<p style="padding: 20px;font-size: 1.25rem; font-family: Segoe UI,Roboto;text-align: center;">';
+        $style_end = '</p></div>';
+        die($style_block . $style_text . $message . $style_end);
+    }
+
+    /**
+     * Return the user's display name as received from Sandstorm
+     *
+     * @return string
+     */
+    public function getDisplayName() {
+        return $this->displayName;
+    }
+
+    /**
+     * Return an array of the user's ARO groups using the Sandstorm permissions.
+     *
+     * Because Sandstorm's permission system is less capable than Open-EMR's,
+     * we pass roles via the permissions header.  Roles include:
+     *
+     *   - role:administrator -> 'Administrators'
+     *   - role:manager
+     *   - role:back office
+     *   - role:front office -> 'Front Office'
+     *   - role:clinician -> 'Clinicians'
+     * @return string[]
+     */
+    public function getAroGroupsWithSandstormPermissions() {
+        $result = array();
+        foreach ($this->permissions as $permission) {
+            if ($permission === 'role:administrator') {
+                array_push($result, 'Administrators');
+            } else if ($permission === 'role:manager') {
+            } else if ($permission === 'role:back office') {
+            } else if ($permission === 'role:front office') {
+                array_push($result, 'Front Office');
+            } else if ($permission === 'role:clinician') {
+                array_push($result, 'Clinicians');
+            }
+        }
+        return $result;
+    }
+
+    /**
+     * Return the user's permissions as received from Sandstorm
+     *
+     * @return string[]
+     */
+    public function getPermissions() {
+        return $this->permissions;
+    }
+
+    /**
+     * Return the URL of the user's picture as received from Sandstorm
+     *
+     * @return string
+     */
+    public function getPictureUrl() {
+        return $this->pictureUrl;
+    }
+
+    /**
+     * Return the user's preferred handle as received from Sandstorm
+     *
+     * @return string
+     */
+    public function getPreferredHandle() {
+        return $this->preferredHandle;
+    }
+
+    /**
+     * Return the user's pronouns as received from Sandstorm
+     *
+     * @return string
+     */
+    public function getPronouns() {
+        return $this->pronouns;
+    }
+
+    /**
+     * Return the session type as received from Sandstorm
+     *
+     * @return string
+     */
+    public function getSessionType() {
+        return $this->sessionType;
+    }
+
+    /**
+     * Return the session ID as received from Sandstorm
+     *
+     * @return string
+     */
+    public function getSessionId() {
+        return $this->sessionId;
+    }
+
+    /**
+     * Return the tab ID as received from Sandstorm
+     *
+     * @return string
+     */
+    public function getTabId() {
+        return $this->tabId;
+    }
+
+    /**
+     * Return the user ID as a binary string
+     *
+     * @return string
+     */
+    public function getUserIdAsBinary() {
+        return $this->userIdBinary;
+    }
+    /**
+     * Return the user ID as received from Sandstorm
+     *
+     * @return string
+     */
+    public function getUserIdAsString() {
+        return $this->userIdString;
+    }
+
+    private function populateWithHeaders() {
+        $headers = getallheaders();
+
+        // displayName
+        $displayNameEncoded = $headers['X-Sandstorm-Username'];
+        if (!empty($displayNameEncoded)) {
+            $this->displayName = urldecode($displayNameEncoded);
+        } else {
+            $this->displayName = '';
+        }
+
+        // permissions
+        $permissionsString = $headers['X-Sandstorm-Permissions'];
+        if (!empty($permissionsString)) {
+            $this->permissions = explode(',', $permissionsString);
+        } else {
+            $this->permissions = array(); 
+        }
+
+        $this->pictureUrl = $headers['X-Sandstorm-User-Picture'];
+        $this->preferredHandle = $headers['X-Sandstorm-Preferred-Handle'];
+        $this->pronouns = $headers['X-Sandstorm-User-Pronouns'];
+        $this->sessionType = $headers['X-Sandstorm-Session-Type'];
+        $this->sessionId = $headers['X-Sandstorm-Session-Id'];
+        $this->tabId = $headers['X-Sandstorm-Tab-Id'];
+        $this->userIdBinary = hex2bin($headers['X-Sandstorm-User-Id']);
+        $this->userIdString = $headers['X-Sandstorm-User-Id'];
+    }
+}
