--- /opt/openemr-7.0.3/openemr/interface/login/login.php.orig	2025-03-23 06:57:50.000000000 +0000
+++ /opt/openemr-7.0.3/openemr/interface/login/login.php	2025-11-04 20:02:48.071077361 +0000
@@ -25,9 +25,13 @@
  * @license   https://github.com/openemr/openemr/blob/master/LICENSE GNU General Public License 3
  */

+$isSandstorm = !!getenv('SANDSTORM');
+
 // prevent UI redressing
-Header("X-Frame-Options: DENY");
-Header("Content-Security-Policy: frame-ancestors 'none'");
+if (!$isSandstorm) {
+    Header("X-Frame-Options: DENY");
+    Header("Content-Security-Policy: frame-ancestors 'none'");
+}

 use OpenEMR\Common\Twig\TwigContainer;
 use OpenEMR\Events\Core\TemplatePageEvent;
@@ -38,7 +42,8 @@
 use Twig\Error\RuntimeError;
 use Twig\Error\SyntaxError;
 
-$ignoreAuth = true;
+$ignoreAuth = !$isSandstorm;
+
 // Set $sessionAllowWrite to true to prevent session concurrency issues during authorization related code
 $sessionAllowWrite = true;
 require_once("../globals.php");
